### ISSUE TITLE
Optimize an async method that ends "return await e" to be non-async "return e"

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#1981]
this optimization would improve the performance of an async method whose only await appears at the end. Not when producing debug code and not when nested in a try block.

@jaredpar @VSadov

**Comment by VSadov** [comment]
This is basically the tail-call optimization for async methods. 

@pharring

**Comment by svick** [comment]
Won't this change the semantics when `e` itself throws an exception (instead of returning a faulted `Task`)? Specifically, `return e` bubbles the exception up the call stack, while `return await e` puts it into the returned `Task`.

**Comment by jaredpar** [comment]
@svick yes it would change the call stack for exceptions of this nature.

**Comment by gafter** [comment]
@svick Given the proposed constraints, the callers will just bubble the exception up, so the end result is the same other than the fact that the call stack is different.

**Comment by svick** [comment]
@gafter I don't understand. Consider the following code:

```c#
Task<int> E()
{
    throw new Exception();
}

async Task<int> ReturnAwaitE()
{
    return await E();
}

Task<int> ReturnE()
{
    return E();
}

…

var t = ReturnAwaitE();
await t.ContinueWith(_ => {}); // await t, but don't throw if it's faulted
Console.WriteLine(t.Exception);
```

This code writes out the exception, but if I replace `ReturnAwaitE()`, with `ReturnE()`, it throws instead; so the end result is very different here.

Doesn't `ReturnAwaitE` fulfill your constraints (the method ends with `return await`, there is no other `await`, there is no `try`)? And isn't `ReturnE` how the optimized version would look like?

**Comment by jaredpar** [comment]
@svick this feature would not change exception semantics in that way.  To do so would turn an optimization into a breaking change.  The implementation of this would have to take exceptions into account when generating the code.  

Even in this case though the compiler can generate much more efficient code than the state machine that is generated today:

``` csharp
Task<int> ReturnE()
{
  try {
    return E();
  } catch  (Exception e) {
    return Task<int>.FromException(e);
 }
}
```

**Comment by pharring** [comment]
:+1: on the idea.
Note that the transform is valid even in the presence of ```.ConfigureAwait(false)```
I had thought of doing this via an analyzer.
I think, in many cases, the extra try/catch is unnecessary. If an exception is thrown from an "await"-able method (instead of returning a faulted Task), then doesn't this indicate an exception in the task handling machinery? That's the sort of exception I'd be happy to let go up to the caller.

**Comment by svick** [comment]
@jaredpar Yeah, that breaking change was what I was afraid of. It wasn't clear to me from the initial description that this was what the optimization would do.

**Comment by jaredpar** [comment]
@pharring if there is another `await` in the function then the existing state machine code will handle it.  If the `return await` call though is the only `await` then the compiler needs to generate the extra try / catch in many cases.  Otherwise it changes the observable behavior of the method.

**Comment by pharring** [comment]
BTW, ```Task.FromException``` is an internal method.

**Comment by stephentoub** [comment]
@pharring, it was in 4.5, it isn't anymore.

**Comment by stephentoub** [comment]
ps Note, though, that the try/catch will also need to specialize catching an OperationCanceledException and returning a canceled task for that case, as that's what the async machinery does today.

**Comment by stephentoub** [comment]
pps And for it to actually remain semantically equivalent, it'd have to appropriately handle ExecutionContext.  In particular, inside of an async method, changes made to ExecutionContext (e.g. putting something into the logical call context) will not be visible to the caller of the async method upon the return of the synchronous invocation.

**Comment by pharring** [comment]
This was proposed a [couple of times already](https://github.com/code-cracker/code-cracker/issues/151)

**Comment by khellang** [comment]
:+1: on this optimization!

**Comment by GSPP** [comment]
Is the C# team willing to take a dependency on the internals of TaskAwaiter here? Previously, that class was opaque and there wasn't any compiler knowledge about it built-in. From the languages point of view that class could do anything at all when awaited.

**Comment by yaakov-h** [comment]
Won't this remove the method in question from an exception stack trace?

Is there a way that this optimisation could be achieved whilst still preserving the stack trace?

**Comment by AlexRadch** [comment]
> [@svick](https://github.com/svick) this feature would not change exception semantics in that way. To do so would turn an optimization into a breaking change. The implementation of this would have to take exceptions into account when generating the code.
> 
> Even in this case though the compiler can generate much more efficient code than the state machine that is generated today:
> 

This optimization may change the thread on which the synchronized code is executed.
Which is very rare, but can lead to an unwanted long-term blocking of the current thread.

```csharp
Task<int> ReturnE()
{
  try {
    Thread.Sleep(30 * 1000); // Long operation in synchronous code
    return E();
  } catch  (Exception e) {
    return Task<int>.FromException(e);
  }
}
```

**Comment by binki** [comment]
@VSadov
> This is basically the tail-call optimization for async methods.
> 
> [@pharring](https://github.com/pharring)

One common use case for `async` methods is to perform a widening cast. For example, the following could not benefit from this optimization:

```csharp
Task<int> CalculateNumberAsync()
{
    return Task.FromResult(4);
}

async Task<long> CalculateLongNumberAsync()
{
    return await CalculateNumberAsync();
}
```

It will be unintuitive if the above code does not get the tail-call optimization while other code does.

@yaakov-h 
> Won't this remove the method in question from an exception stack trace?
> 
> Is there a way that this optimisation could be achieved whilst still preserving the stack trace?

This is a concern I share. However, I think I am incorrect to hold this as a concern.

If I recall correctly, synchronous methods can be removed from the call stack when they are inlined. So if another optimization already removes frames from the call stack in non-Debug mode, then the exact stack length is not something that should be relied upon. The async machinery itself already **adds** a bunch of extra stack frames and I assume that those are supposed to be treated like an implementation detail anyway.
